### ISSUE TITLE
TLS configuration inconsistencies and complexity for TCP versus RELP transports 

### ISSUE THREAD

**Issue by JPvRiel** [rsyslog/rsyslog#435]
Given a central server that needs to support both clients that are able to use secure and non-secure log forwarding, the following issues are noted:
- rsyslog is able to run more than one TCP input on different ports, with the TLS configuration (StreamDriver.mode) managed per port, but it's cumbersome to configure.
- current rainerscript configuration styles allow for default global stream driver configuration directives or specification during  imtcp module loading, but not when defining imtcp inputs.
- With RELP, it is possible to configure TLS per RELP imrelp input after loading the imrelp module.
- The way in which TLS is configured for TCP is more tedious, cumbersome and inconsistent with the way it is configured for RELP.
- However, RELP doesn't support 'certvalid' auth mode like TCP, although that could be implied using tls.permittedPeer="*"?
  (Tested on Rsyslog 8.11 with CentOS 7)

Suggested fixes
- Unify and make configuration directives for TLS consistant between input and output modules
- Allow global config directives for TLS that can apply to TLS for both RELP and TCP
- Allow all inputs that can support TLS to overide/supply TLS parameters (don't force config to be set within the module load directive)
- Allow RELP to specify certvalid auth mode

Example config /etc/rsyslog.d/remote_in.conf highlights the differences and complexity in configuration:

```
global(
  defaultNetstreamDriver="ptcp" 
  #TCP TLS
  defaultNetstreamDriverCAFile="/etc/pki/ca-trust/source/anchors/SB_CA_rsyslog_test_root.crt"
  defaultNetstreamDriverCertFile="/etc/pki/tls/certs/05766testrsyslog8srv.rsa.crt"
  defaultNetstreamDriverKeyFile="/etc/pki/tls/private/05766testrsyslog8srv.rsa.key"
)

module(load="imptcp")
  # plain tcp without TLS
module(
  load="imtcp"
  streamDriver.name="gtls"
  streamDriver.mode="1"
    #This indicates TLS is enabled. 0 indicated disabled.
  streamDriver.authMode="x509/certvalid"
  #streamDriver.authMode="x509/name"
  #permittedPeer="*.example.com"
)
module(load="imrelp")

# provides TCP syslog reception
input(type="imptcp" port="514" ruleset="remote_tcp_5424")

# provides secure TCP syslog reception
input(
  type="imtcp"
  name="tcp_secure"
  port="6514"
  ruleset="remote_tcp_secure_5424"
)

# proivdes RELP reception
input(type="imrelp" name="imrepl" port="2514" ruleset="remote_relp_5424" tls="off" )

# provides secure RELP reception (without client auth enforced)
input(
  type="imrelp"
  name="imrelp_secure"
  port="8514"
  ruleset="remote_relp_secure_5424"
  tls="on"
  tls.dhbits="2048"
  tls.caCert="/etc/pki/ca-trust/source/anchors/SB_CA_rsyslog_test_root.crt"
  tls.myCert="/etc/pki/tls/certs/05766testrsyslog8srv.rsa.crt"
  tls.myPrivKey="/etc/pki/tls/private/05766testrsyslog8srv.rsa.key"
  tls.authMode="name"
  tls.permittedPeer="*.example.com"
)

# Note each input rulset has an action to relay logs to a localhost port in 2xxxx range for live testing. If port closed, direct queue implies logs will be dropped. E.g. one can use netcat to see relayed output.

RuleSet (name="remote_tcp_5424")
{ action (
    name="test_relay_tcp_5424" 
    type="omfwd"
    target="127.0.0.1"
    port="20514"
    protocol="tcp"
    queue.type="Direct"
  )
}

RuleSet (name="remote_tcp_secure_5424")
{ action (
    name="test_relay_tcp_secure_5424" 
    type="omfwd"
    target="127.0.0.1"
    port="26514"
    protocol="tcp"
    queue.type="Direct"
  )
}

RuleSet (name="remote_relp_5424")
{ action (
    name="test_relay_relp_5424" 
    type="omfwd"
    target="127.0.0.1"
    port="22514"
    protocol="tcp"
    queue.type="Direct"
  )
}

RuleSet (name="remote_relp_secure_5424")
{ action (
    name="test_relay_relp_5424_secure" 
    type="omfwd"
    target="127.0.0.1"
    port="28514"
    protocol="tcp"
    queue.type="Direct"
  )
}
```

**Comment by JPvRiel** [comment]
To add another note, I'm having issues trying to run RELP without requiring the client to provide a peer certificate (e.g. just normal server authentication without client authentication). The error is:

```
rsyslogd: imrelp[7514]: authentication error 'peer did not provide a certificate', peer is '' [v8.29.0 try http://www.rsyslog.com/e/2353 ]
```

I tried both not specifying `tls.authMode` (hoping it would default to none) and `tls.authMode="anon"` or `tls.authMode="none"`, but it seems gnuTLS with imrelp insists, by default, on requesting a client certificate. When I run `openssl s_client` tests, I can see the syslog RELP endpoint requesting a client cert.

Trying `tls.authMode="anon"` with RELP simply gets this error, so it's not accepted (and inconsistent with the tcp stream driver settings for imtcp):

```
rsyslogd: imrelp: invalid auth mode 'anon' [v8.29.0 try http://www.rsyslog.com/e/2291 ]
```

However, 'anon' is valid mode for the TLS and TCP (streamDriver) config. In `nsd_gtls.h`

```
	enum { 
		GTLS_AUTH_CERTNAME = 0,
		GTLS_AUTH_CERTFINGERPRINT = 1,
		GTLS_AUTH_CERTVALID = 2,
		GTLS_AUTH_CERTANON = 3
	} authMode;
```

But in `librelp.h`, it does seem to have a none mode, but it's unclear how to set it.

```
/** authentication modes for TLS */
typedef enum relpAuthMode_e {
	eRelpAuthMode_None = 0,
	eRelpAuthMode_Fingerprint = 1,
	eRelpAuthMode_Name = 2
} relpAuthMode_t;
```

I tried setting guessing `none`, but also got `invalid auth mode 'none'`. 

The `relpSrvSetAuthMode` function accepts a Null value and probably doesn't set it then, but it's not clear how to get config to leave this as NULL or call it with NULL, given not setting `tls.authMode` still caused the peer cert error.

Also, as per the `gnutls_certificate_server_set_request` function documentation:

> If you do not call this function then the client will not be asked to send a certificate.

So the suggested enhancement would be to have a switch statement which avoids calling the function for imrelp if the user wishes to have only server authentication.

While we all know it's better to enforce client authentication, the reality is that some system owners wont have properly provision client certificates, so I'd rather collect logs in a half secure way (encryption and sever authN) than force clients to choose a less secure option because they couldn't do client auth (e.g. not encrypted, etc)
